Add capabilities objects

dvarrazzo opened the pull request:
The object, exposed as `psycopg.capabilities`, exposes `has_STUFF()` methods to verify if a certain feature is supported.

If `check=True` is specified, raise an exception instead of returning False, with a finely crafted message such as:

> the feature 'Connection.pipeline()' is not available: you are using a psycopg[c] libpq wrapper built with libpq version 9.5.3; the feature requires libpq version at least 14.0
>
> the feature 'Connection.pipeline()' is not available: the client libpq version (imported from system libraries) is 13.4; the feature requires libpq version at least 14.0
>
> the feature 'Connection.pipeline()' is not available: the client libpq version (imported from psycopg[binary] version 3.2.0) is 13.4; the feature requires libpq version at least 14.0

helping to diagnose the problem.

As a side effect, added a `pq.version_pretty()` function to convert the numieric version to the postgres version. Code refactored to use it instead of homegrown `// 10000`, divmod, and similar.

The `Pipeline.is_supported()` method has been reimplemented using `Capabilities.has_pipeline()`.

Close #772

asqui commented:
Drive-by naming nitpick: Would `raise: bool` be a better name for `check`?

I think `capabilities.has_pipeline(raise=True)` is a bit more descriptive about the behaviour to someone who is unfamiliar with the capabilities API.

asqui commented:
(PS: I agree that the `has_STUFF()` API is nicer than `has_capability(capability: str)`, and latter gives extra flexibility that I think is not necessary. 👍 )

dvarrazzo commented:
Hi, nice to see you dropping by

> Drive-by naming nitpick: Would `raise: bool` be a better name for `check`?
>
> I think `capabilities.has_pipeline(raise=True)` is a bit more descriptive about the behaviour to someone who is unfamiliar with the capabilities API.

You might be right. I should "check" what other use we make of similarly named parameter around in the code. Actually, one that I have in mind is the pool which has a `check` method and build parameter... and they mean a different thing.

So, at a glance, raise might be a better name.

dlax commented on psycopg/psycopg/_capabilities.py:
```diff
@@ -0,0 +1,89 @@
+
+    def has_hostaddr(self, check: bool = False) -> bool:
+        """Check if the `~ConnectionInfo.hostaddr` attribute is implemented."""
+        return self._has_feature("Connection.pipeline()", 120000, check=check)
```
Should be `"ConnectionInfo.hostaddr"`.

dlax commented on docs/api/connections.rst:
```diff
@@ -287,6 +287,12 @@ The `!Connection` class
+        .. note::
+
+            You can use the `~Capabilities.has_cancel_safe` capability to check
+            if `!cancel_safe()` will will not fall back on the legacy
```
typo: will will

dvarrazzo replied:
Fixed, thank you

dlax commented on docs/api/objects.rst:
```diff
@@ -109,6 +110,58 @@ Connection information
+    the methods return a boolean value stating if the capability is supported,
+    which can be used by a program to degrade gracefully::
+
+        if psycopg.capabilities.has_pipeline(check=True)
```
`check=False` (or omitted) in this case, no?

dvarrazzo replied:
Yes, it's a mistake. Fixed, thank you

dvarrazzo commented on docs/news.rst:
```diff
@@ -19,6 +19,8 @@ Psycopg 3.2 (unreleased)
 - Add :ref:`raw-query-cursors` to execute queries using placeholders in
   PostgreSQL format (`$1`, `$2`...) (:ticket:`#560`).
+- Add `psycopg.capabilities` object to :ref:`inspect the libpq capabilities
+  <capabilities>` (:ticket:`#772`).
```
The `_` syntax is a reST feature and works only in the same file. Because of this limitation, Sphinx added the `ref` role which allows linking to references in different files in the doc.

dvarrazzo commented on psycopg/psycopg/_capabilities.py:
```diff
@@ -11,36 +11,49 @@
+
+        The feature requires libpq 12.0 and greater.
+        """
+        return self._has_feature("Connection.info.hostaddr", 120000, check=check)
```
Yes, it was an early pasto.

asqui commented:
Ok, I was actually thinking of `raises: bool` which I'd seen in another library, but I think @dlax's citation from `subprocess.run(..., check=True)` is a more compelling argument for calling it `check`, so I withdraw my suggestion 😄

dlax commented on psycopg/psycopg/_pipeline.py:
```diff
@@ -63,36 +64,12 @@ def status(self) -> pq.PipelineStatus:
-                f"pipeline mode not supported: {self._not_supported_reason()}"
-            )
+        if not self._is_supported:
+            capabilities.has_pipeline(check=True)
```
IIUC, we'd now always run `capabilities.has_pipeline(check=True)` when entering the pipeline, unless `is_supported()` has been called before to set the `_is_supported` attribute, no?

Also, it might interesting to only call `capabilities.has_pipeline()` once; something like 663e64265cb9aa422aa401e03877ab063f6ef08f.

dvarrazzo replied:
Yes, that's xorrect. The if is a micro optimization to avoid to call `has_pipeline()` every time in the most common path (when we already established that pipelines are supported).

Maybe it's worth nothing, or maybe in general I should have added caching in the Capabilities object, if my gripe is to avoid to call `pq.version()` over and over. 

TBH I've been a bit lazy here and I didn't add caching because I thought it'd make tests bit more cumbersome... But maybe I should do that.

dvarrazzo commented:
I have added caching on the capabilities instance, and simplified the pipeline objects as a consequence:

https://github.com/psycopg/psycopg/blob/c4dc5d1ce236ded8a31bf25f8fed8791901f372c/psycopg/psycopg/_pipeline.py#L63-L68

If that's ok, can merge this, then use it in:

https://github.com/psycopg/psycopg/blob/7cf456a06a79b08bdbd825913fa1fe998eb71044/psycopg/psycopg/connection.py#L277-L278

and merge #785 too.

dlax commented:
Out of curiosity, why not using `functools.cache()`?

dvarrazzo commented:
> Out of curiosity, why not using `functools.cache()`?

It doesn't declare explicitly that it works to wrap a method. For instance I don't know if and how it includes the `self` parameter in the key calculation. Therefore I prefer to use `@cache` only on pure functions, not methods.
